File: ingest/build_corpus.py
#!/usr/bin/env python3
"""Build JSONL chunks for viz-rag embedding. Usage: python ingest/build_corpus.py"""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SOURCES = ROOT / "sources"
OUT = ROOT / "chunks" / "corpus.jsonl"


def split_sections(text: str) -> list[tuple[str, str]]:
    parts = re.split(r"\n(?=## )", text)
    out: list[tuple[str, str]] = []
    for part in parts:
        lines = part.splitlines()
        if lines and lines[0].startswith("## "):
            out.append((lines[0][3:].strip(), "\n".join(lines[1:]).strip()))
        else:
            out.append(("", part.strip()))
    return out


def strip_yaml_rmd(text: str) -> str:
    if text.startswith("---"):
        end = text.find("\n---\n", 3)
        if end != -1:
            return text[end + 5 :]
    return text


def strip_html(html: str) -> str:
    html = re.sub(r"<script[^>]*>[\s\S]*?</script>", " ", html, flags=re.I)
    html = re.sub(r"<style[^>]*>[\s\S]*?</style>", " ", html, flags=re.I)
    html = re.sub(r"<[^>]+>", " ", html)
    html = html.replace("&nbsp;", " ").replace("&amp;", "&")
    return re.sub(r"\s+", " ", html).strip()


def write_chunk(
    fh,
    source: str,
    path: str,
    section: str,
    text: str,
    meta: dict,
    counter: list[int],
) -> None:
    text = text.strip()
    if not text:
        return
    counter[0] += 1
    row = {
        "id": str(counter[0]),
        "source": source,
        "path": path,
        "section": section,
        "text": text,
        **meta,
    }
    fh.write(json.dumps(row, ensure_ascii=False) + "\n")


def ingest_rmd(fh, path: Path, source: str, meta: dict, counter: list[int]) -> None:
    text = strip_yaml_rmd(path.read_text(encoding="utf-8", errors="replace"))
    rel = str(path.relative_to(ROOT)).replace("\\", "/")
    for heading, body in split_sections(text):
        section = heading or path.name
        write_chunk(fh, source, rel, section, body, meta, counter)


def ingest_md(fh, path: Path, source: str, meta: dict, counter: list[int]) -> None:
    text = path.read_text(encoding="utf-8", errors="replace")
    rel = str(path.relative_to(ROOT)).replace("\\", "/")
    for heading, body in split_sections(text):
        section = heading or path.name
        write_chunk(fh, source, rel, section, body, meta, counter)


def ingest_html_caveat(fh, path: Path, source: str, meta: dict, counter: list[int]) -> None:
    html = path.read_text(encoding="utf-8", errors="replace")
    m = re.search(r"<title>([^<]+)</title>", html, re.I)
    title = m.group(1).strip() if m else path.name
    rel = str(path.relative_to(ROOT)).replace("\\", "/")
    write_chunk(fh, source, rel, title, strip_html(html), meta, counter)


def ingest_r_file(fh, path: Path, source: str, meta: dict, counter: list[int]) -> None:
    try:
        rel = str(path.relative_to(ROOT)).replace("\\", "/")
    except ValueError:
        rel = str(path).replace("\\", "/")
    write_chunk(fh, source, rel, path.name, path.read_text(encoding="utf-8", errors="replace"), meta, counter)


def main() -> None:
    OUT.parent.mkdir(parents=True, exist_ok=True)
    counter = [0]

    with OUT.open("w", encoding="utf-8") as fh:
        dtv = SOURCES / "data_to_viz"
        if dtv.is_dir():
            story = dtv / "story"
            for rmd in sorted(story.glob("*.Rmd")):
                ingest_rmd(
                    fh,
                    rmd,
                    "data_to_viz",
                    {"language": "r", "role": "chart_selection", "package": "ggplot2"},
                    counter,
                )
            readme = dtv / "Example_dataset" / "Readme.md"
            if readme.is_file():
                ingest_md(
                    fh,
                    readme,
                    "data_to_viz",
                    {"language": "r", "role": "data_shape", "package": "ggplot2"},
                    counter,
                )
            caveat = dtv / "caveat"
            if caveat.is_dir():
                for html in sorted(caveat.glob("*.html")):
                    ingest_html_caveat(
                        fh, html, "data_to_viz", {"language": "en", "role": "caveat"}, counter
                    )
        else:
            print("Missing sources/data_to_viz — run scripts/fetch_sources.sh")

        ggp = SOURCES / "ggplot2" / "vignettes"
        if ggp.is_dir():
            for vignette in sorted(ggp.glob("*.qmd")) + sorted(ggp.glob("*.Rmd")):
                ingest_rmd(
                    fh,
                    vignette,
                    "ggplot2",
                    {"language": "r", "role": "implementation", "package": "ggplot2"},
                    counter,
                )
        else:
            print("Missing sources/ggplot2 — run scripts/fetch_sources.sh")

        for sub, role in (
            ("corpus/style", "style"),
            ("corpus/notes", "theory"),
            ("corpus/lessons", "lesson"),
        ):
            d = ROOT / sub
            if d.is_dir():
                for md in sorted(d.glob("*.md")):
                    ingest_md(
                        fh,
                        md,
                        "local",
                        {"language": "en", "role": role, "package": "ggplot2"},
                        counter,
                    )

        tt_lessons = ROOT.parent / "tidytuesday" / "VIZ_LESSONS.md"
        if tt_lessons.is_file():
            ingest_md(
                fh,
                tt_lessons,
                "tidytuesday",
                {"language": "en", "role": "lesson", "package": "ggplot2"},
                counter,
            )

        tt_r = ROOT.parent / "tidytuesday" / "2026_06_16_uk_baby_names" / "R"
        if tt_r.is_dir():
            for rfile in sorted(tt_r.glob("*.R")):
                if rfile.name == "paths.R":
                    continue
                ingest_r_file(
                    fh,
                    rfile,
                    "tidytuesday",
                    {"language": "r", "role": "recipe", "package": "ggplot2"},
                    counter,
                )

    print(f"Wrote {counter[0]} chunks to {OUT}")


if __name__ == "__main__":
    main()

File: ingest/build_corpus.R
# Build JSONL chunk files for embedding / RAG retrieval.
# Usage: source("ingest/build_corpus.R"); build_corpus()

library(jsonlite)

split_sections <- function(text) {
  parts <- strsplit(text, "\n(?=## )", perl = TRUE)[[1]]
  if (length(parts) == 1L) {
    return(list(list(heading = "", body = text)))
  }
  lapply(parts, function(part) {
    lines <- strsplit(part, "\n", fixed = TRUE)[[1]]
    heading <- if (grepl("^## ", lines[1])) sub("^## ", "", lines[1]) else ""
    body <- paste(if (heading == "") lines else lines[-1], collapse = "\n")
    list(heading = heading, body = trimws(body))
  })
}

write_chunk <- function(chunks, con, source, path, section, text, meta = list()) {
  if (!nzchar(trimws(text))) {
    return(invisible(NULL))
  }
  row <- c(
    list(
      id = paste0(length(chunks) + 1L),
      source = source,
      path = path,
      section = section,
      text = text
    ),
    meta
  )
  cat(toJSON(row, auto_unbox = TRUE), "\n", file = con, append = TRUE)
  invisible(length(chunks) + 1L)
}

ingest_rmd <- function(path, source, meta, con, chunks_len) {
  text <- paste(readLines(path, warn = FALSE), collapse = "\n")
  yaml_end <- regexpr("\n---\n", text, fixed = TRUE)[1]
  if (yaml_end > 0) {
    text <- substr(text, yaml_end + 5, nchar(text))
  }
  sections <- split_sections(text)
  rel <- path
  for (sec in sections) {
    label <- if (nzchar(sec$heading)) sec$heading else basename(path)
    write_chunk(
      list(), con, source, rel, label, sec$body, meta
    )
  }
}

ingest_markdown <- function(path, source, meta, con) {
  text <- paste(readLines(path, warn = FALSE), collapse = "\n")
  sections <- split_sections(text)
  rel <- path
  for (sec in sections) {
    label <- if (nzchar(sec$heading)) sec$heading else basename(path)
    write_chunk(list(), con, source, rel, label, sec$body, meta)
  }
}

strip_html <- function(html) {
  html <- gsub("<script[^>]*>[\\s\\S]*?</script>", " ", html, perl = TRUE, ignore.case = TRUE)
  html <- gsub("<style[^>]*>[\\s\\S]*?</style>", " ", html, perl = TRUE, ignore.case = TRUE)
  html <- gsub("<[^>]+>", " ", html)
  html <- gsub("&nbsp;", " ", html, fixed = TRUE)
  html <- gsub("&amp;", "&", html, fixed = TRUE)
  html <- gsub("\\s+", " ", html)
  trimws(html)
}

ingest_html_caveat <- function(path, source, meta, con) {
  html <- paste(readLines(path, warn = FALSE), collapse = "\n")
  title <- sub(".*<title>([^<]+)</title>.*", "\\1", html)
  if (identical(title, html)) title <- basename(path)
  body <- strip_html(html)
  write_chunk(list(), con, source, path, title, body, meta)
}

ingest_r_file <- function(path, source, meta, con) {
  text <- paste(readLines(path, warn = FALSE), collapse = "\n")
  rel <- tryCatch(
    sub(paste0("^", normalizePath(project_root, winslash = "/"), "/?"), "", normalizePath(path, winslash = "/")),
    error = function(e) path
  )
  write_chunk(list(), con, source, rel, basename(path), text, meta)
}

build_corpus <- function(project_root = ".") {
  project_root <- normalizePath(project_root, winslash = "/", mustWork = FALSE)
  sources <- file.path(project_root, "sources")
  out_dir <- file.path(project_root, "chunks")
  dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
  out_file <- file.path(out_dir, "corpus.jsonl")
  if (file.exists(out_file)) file.remove(out_file)
  file.create(out_file)

  # data_to_viz stories
  dtv <- file.path(sources, "data_to_viz")
  if (dir.exists(dtv)) {
    rmds <- list.files(file.path(dtv, "story"), pattern = "\\.Rmd$", full.names = TRUE)
    for (f in rmds) {
      ingest_rmd(
        f, "data_to_viz",
        list(language = "r", role = "chart_selection", package = "ggplot2"),
        out_file, 0
      )
    }
    readme <- file.path(dtv, "Example_dataset", "Readme.md")
    if (file.exists(readme)) {
      ingest_markdown(
        readme, "data_to_viz",
        list(language = "r", role = "data_shape", package = "ggplot2"),
        out_file
      )
    }
    caveat_dir <- file.path(dtv, "caveat")
    if (dir.exists(caveat_dir)) {
      htmls <- list.files(caveat_dir, pattern = "\\.html$", full.names = TRUE)
      for (f in htmls) {
        ingest_html_caveat(
          f, "data_to_viz",
          list(language = "en", role = "caveat"),
          out_file
        )
      }
    }
  } else {
    message("Missing sources/data_to_viz — run scripts/fetch_sources.sh first")
  }

  # ggplot2 vignettes
  ggp <- file.path(sources, "ggplot2", "vignettes")
  if (dir.exists(ggp)) {
  vignettes <- list.files(ggp, pattern = "\\.(qmd|Rmd)$", full.names = TRUE)
    for (f in vignettes) {
      ingest_rmd(
        f, "ggplot2",
        list(language = "r", role = "implementation", package = "ggplot2"),
        out_file, 0
      )
    }
  } else {
    message("Missing sources/ggplot2 — run scripts/fetch_sources.sh first")
  }

  # Local curated markdown
  for (subdir in c("corpus/style", "corpus/notes")) {
    dir_path <- file.path(project_root, subdir)
    if (!dir.exists(dir_path)) next
    mds <- list.files(dir_path, pattern = "\\.md$", full.names = TRUE)
    for (f in mds) {
      role <- if (grepl("style", subdir)) "style" else "theory"
      ingest_markdown(
        f, "local",
        list(language = "en", role = role, package = "ggplot2"),
        out_file
      )
    }
  }

  # Optional: personal R recipes from tidytuesday
  tt_r <- normalizePath(file.path(project_root, "../tidytuesday/2026_06_16_uk_baby_names/R"), winslash = "/", mustWork = FALSE)
  if (dir.exists(tt_r)) {
    r_files <- list.files(tt_r, pattern = "\\.R$", full.names = TRUE)
    r_files <- r_files[!grepl("paths\\.R$", r_files)]
    for (f in r_files) {
      ingest_r_file(
        f, "tidytuesday",
        list(language = "r", role = "recipe", package = "ggplot2"),
        out_file
      )
    }
  }

  n <- length(readLines(out_file))
  message("Wrote ", n, " chunks to ", out_file)
  invisible(out_file)
}
